### PR TITLE
Improvements the mouse and keys of the MenuBar. 

### PR DESCRIPTION
**Pull request by BDisp**
The hot keys were been called to the wrong Toplevel even it isn't the current. Mouse changes was to get a better response. And there was fixed some minor bugs due the various previous interventions.

### CODE REVIEW

**Comment by BDisp** [comment]
I think this https://github.com/migueldeicaza/gui.cs/pull/596/commits/ce30a837642cc72aef6433e9322a6c4970c13028 fixes #436.

**Comment by migueldeicaza** [comment]
We probably need a better description of what this change does on the description, it is hard to judge the changes without it.

**Comment by BDisp** [comment]
Without merge with https://github.com/migueldeicaza/gui.cs/pull/596/commits/4a1f5a9309cf10d45d13c3b7d87ba7ceb8303a6a `Alt `key doesn't work in the `Menu`.

**Comment by tig** [comment]
@BDisp I'm very confused about what this PR does and doesn't do. 

It makes a lot of changes that are hard to follow in the diff. Can you try to explain the problem and the fix more clearly?

**Comment by BDisp** [comment]
> 
> 
> @BDisp I'm very confused about what this PR does and doesn't do.
> 
> It makes a lot of changes that are hard to follow in the diff. Can you try to explain the problem and the fix more clearly?

`OnKeyUp `and `OnKeyDown `in reality I reset them for what remains the same because my intention was to activate the `Menu `right after pressing and not when releasing the key but this is not effective because it activates right when switching between windows and leaves the `Menu `activated no need. At the mouse level, I activated both `ButtonPressed `and `ButtonDoubleClicked `to open and close the `Menu `more quickly. Now it works alternately, that is, if the `Menu `is closed both events will open and close it, otherwise. There is another commit that also solves a problem in which if you open the "`Text Editor Demo`" in the `Example `project, pressing` Ctrl + Q `would trigger the `Application.Top` Quit method that does not belong to the `Toplevel `current.

**Comment by BDisp** [comment]
@tig here is a video explaining what I was meaning about the Quit that belongs to the `Application.Top` is running in the `Text Editor Demo`. If I choose "Ok" to quit after I close the editor the application exits.

toplevel-keys

File: Terminal.Gui/ConsoleDrivers/WindowsDriver.cs [hunk 1]
```diff
@@ -892,6 +879,32 @@ MouseEvent ToDriverMouse (WindowsConsole.MouseEventRecord mouseEvent)
 			};
 		}
 
+		private async Task ProcessButtonDoubleClickedAsync ()
+		{
+			await Task.Delay (200);
+			IsButtonDoubleClicked = false;
+		}
+
+		async Task ProcessContinuousButtonPressedAsync (WindowsConsole.MouseEventRecord mouseEvent, MouseFlags mouseFlag)
```
Comment (tig) [inline Terminal.Gui/ConsoleDrivers/WindowsDriver.cs:888]: I like this pattern. It makes a lot of sense to me, now that I've studied it.

File: Terminal.Gui/ConsoleDrivers/WindowsDriver.cs [hunk 2]
```diff
@@ -892,6 +879,32 @@ MouseEvent ToDriverMouse (WindowsConsole.MouseEventRecord mouseEvent)
 			};
 		}
 
+		private async Task ProcessButtonDoubleClickedAsync ()
```
Comment (tig) [inline Terminal.Gui/ConsoleDrivers/WindowsDriver.cs:882]: Coding style says "don't use `private`" as it's the default and not needed.

File: Terminal.Gui/ConsoleDrivers/WindowsDriver.cs [hunk 3]
```diff
@@ -695,7 +695,8 @@ void IMainLoopDriver.MainIteration ()
 						keyUpHandler (new KeyEvent (map, keyModifiers));
 					}
 				}
-				keyModifiers = null;
+				if (!inputEvent.KeyEvent.bKeyDown)
+					keyModifiers = null;
```
Comment (tig) [inline Terminal.Gui/ConsoleDrivers/WindowsDriver.cs:699]: Please wrap all if's with `{` and `}`.

File: Terminal.Gui/ConsoleDrivers/WindowsDriver.cs [hunk 4]
```diff
@@ -892,6 +879,32 @@ MouseEvent ToDriverMouse (WindowsConsole.MouseEventRecord mouseEvent)
...
+			while (IsButtonPressed) {
+				await Task.Delay (200);
+				var me = new MouseEvent () {
+					X = mouseEvent.MousePosition.X,
+					Y = mouseEvent.MousePosition.Y,
+					Flags = mouseFlag
+				};
+
+				var view = Application.wantContinuousButtonPressedView;
+				if (view == null)
```
Comment (tig) [inline Terminal.Gui/ConsoleDrivers/WindowsDriver.cs:899]: Wrap `break;` in `{}`

File: Terminal.Gui/ConsoleDrivers/WindowsDriver.cs [hunk 5]
```diff
@@ -892,6 +879,32 @@ MouseEvent ToDriverMouse (WindowsConsole.MouseEventRecord mouseEvent)
...
+					break;
+				if (IsButtonPressed && (mouseFlag & MouseFlags.ReportMousePosition) == 0) {
+					mouseHandler (me);
+					//mainLoop.Driver.Wakeup ();
```
Comment (tig) [inline Terminal.Gui/ConsoleDrivers/WindowsDriver.cs:903]: Can this comment be deleted?

File: Terminal.Gui/Core/Application.cs [hunk 6]
```diff
@@ -352,7 +352,7 @@ static void ProcessMouseEvent (MouseEvent me)
 					View = view
 				};
 				if (OutsideFrame (new Point (nme.X, nme.Y), mouseGrabView.Frame))
-					lastMouseOwnerView.OnMouseLeave (me);
+					lastMouseOwnerView?.OnMouseLeave (me);
```
Comment (tig) [inline Terminal.Gui/Core/Application.cs:355]: Use `{}`.

File: Terminal.Gui/Core/View.cs [hunk 7]
```diff
@@ -989,7 +989,7 @@ public override bool ProcessHotKey (KeyEvent keyEvent)
 			if (subviews == null || subviews.Count == 0)
 				return false;
 			foreach (var view in subviews)
-				if (view.ProcessHotKey (keyEvent))
+				if (view.SuperView.IsCurrentTop && view.ProcessHotKey (keyEvent))
```
Comment (tig) [inline Terminal.Gui/Core/View.cs:992]: This comment is broader than this PR, but anytime I see three functions that have EXACTLY the same logic in them, with a fair amount of logic, I wonder if they could be refactored to share implementation to reduce code and risk of one of them having a bug.

ARE all 3 the same???

File: Terminal.Gui/Views/Menu.cs [hunk 8]
```diff
@@ -815,6 +816,7 @@ internal void CloseMenu (bool reopen = false, bool isSubMenu = false)
 					IsMenuOpen = false;
 					PositionCursor ();
 				}
+				IsMenuOpen = false;
```
Comment (tig) [inline Terminal.Gui/Views/Menu.cs:819]: Both branches of the `if` above set `IsMenuOpen = false`. Why is this needed?

File: Terminal.Gui/Core/View.cs (hunk 7 above)
Comment (BDisp) [reply Terminal.Gui/Core/View.cs:992]: These processes have different behaviours but with the same logic. They were already there and I don't wanna break anything. Later we can see if it will be possible to congregate all of them.

File: Terminal.Gui/Views/Menu.cs (hunk 8 above)
Comment (BDisp) [reply Terminal.Gui/Views/Menu.cs:819]: You're right. My bad :-)

**Comment by tig** [comment]
You are awesome. Thank you!
